### PR TITLE
copier: init at 7.0.1

### PR DESCRIPTION
**Pull request by jonringer**
###### Motivation for this change
Thought I would try my hand at packaging this, someone mentioned they had great difficulty doing so, and I would agree.

https://discourse.nixos.org/t/buildpythonpackage-throwing-dependency-version-errors-when-i-dont-belive-it-should-be/15818

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/copier/default.nix [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  # poetry-dynamic-versioning does not currently support being able
+  # to set the version explicitly as a package maintainer
+  # see: https://github.com/mtkennerly/poetry-dynamic-versioning/issues/62
+  postFixup = ''
+    substituteInPlace $out/${python3.sitePackages}/copier/__init__.py \
+      --replace "0.0.0" "5.1.0+git${src.rev}"
+  '';
+
+  meta = with lib; {
+    description = "Library and command-line utility for rendering projects templates.";
```
Comment (fabaff) [inline pkgs/tools/misc/copier/default.nix:56]:
```suggestion
    description = "Library and command-line utility for rendering projects templates";
```

File: pkgs/tools/misc/copier/default.nix [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    rev = "9f05fdf77d725933be3f779f1c3215467851f92a";
+    sha256 = "sha256-uUM9Zzh2uOEqcwZxubjj4PawHmqZhQDFe7aVQKe6lM8=";
+  };
+
+  postPatch = ''
+    ${gnused}/bin/sed -e '/packaging/d' -i pyproject.toml
+  '';
+
+  nativeBuildInputs = [
+    poetry
```
Comment (fabaff) [inline pkgs/tools/misc/copier/default.nix:20]: Upstream seems to use `poetry-core` as [build backend](https://github.com/copier-org/copier/blob/master/pyproject.toml#L120).

File: pkgs/tools/misc/copier/default.nix [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  # to set the version explicitly as a package maintainer
+  # see: https://github.com/mtkennerly/poetry-dynamic-versioning/issues/62
+  postFixup = ''
+    substituteInPlace $out/${python3.sitePackages}/copier/__init__.py \
+      --replace "0.0.0" "5.1.0+git${src.rev}"
+  '';
+
+  meta = with lib; {
+    description = "Library and command-line utility for rendering projects templates.";
+    homepage = "github.com/copier-org/copier";
```
Comment (fabaff) [inline pkgs/tools/misc/copier/default.nix:57]: Opinionated change

```suggestion
    homepage = "https://copier.readthedocs.io";
```

File: pkgs/development/python-modules/dunamai/default.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+  src = fetchFromGitHub {
+    owner = "mtkennerly";
+    repo = "dunamai";
+    rev = "v${version}";
+    sha256 = "sha256-ocPkklKJTIT9YizQ4HwDv/r+9G+/XPxMYAgf9DuzDto=";
+  };
+
+  nativeBuildInputs = [
+    poetry
```
Comment (fabaff) [inline pkgs/development/python-modules/dunamai/default.nix:24]: Perhaps https://github.com/mtkennerly/dunamai/pull/35 gets merged quickly then we could move to `poetry-core`.

File: pkgs/development/python-modules/jinja2-ansible-filters/default.nix [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    echo ${version} > VERSION
+  '';
+
+  propagatedBuildInputs = [
+    jinja2
+    pyyaml
+  ];
+
+  # no tests include in sdist, and source not available
+  doCheck = false;
```
Comment (fabaff) [inline pkgs/development/python-modules/jinja2-ansible-filters/default.nix:26]: The upstream repo contains a couple of [`tests`](https://gitlab.com/dreamer-labs/libraries/jinja2-ansible-filters/-/tree/master/tests).

File: pkgs/development/python-modules/poetry-dynamic-versioning/default.nix [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    inherit pname version;
+    sha256 = "5c0e7b22560db76812057ef95dadad662ecc63eb270145787eabe73da7c222f9";
+  };
+
+  propagatedBuildInputs = [
+    dunamai
+    jinja2
+    tomlkit
+  ];
+
```
Comment (fabaff) [inline pkgs/development/python-modules/poetry-dynamic-versioning/default.nix:21]: [Tests](https://github.com/mtkennerly/poetry-dynamic-versioning/tree/master/tests) would be available if the source from GitHub is used.

File: pkgs/tools/misc/copier/default.nix [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+  src = fetchFromGitHub {
+    owner = "copier-org";
+    repo = "copier";
+    rev = "9f05fdf77d725933be3f779f1c3215467851f92a";
+    sha256 = "sha256-uUM9Zzh2uOEqcwZxubjj4PawHmqZhQDFe7aVQKe6lM8=";
+  };
+
+  postPatch = ''
+    ${gnused}/bin/sed -e '/packaging/d' -i pyproject.toml
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/copier/default.nix:16]:
```suggestion
    sed -e '/packaging/d' -i pyproject.toml
```

File: pkgs/tools/misc/copier/default.nix [hunk 8]
```diff
@@ -0,0 +1,61 @@
+{ lib, gnused, poetry, git, python3, fetchFromGitHub }:
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/copier/default.nix:1]:
```suggestion
{ lib, poetry, git, python3, fetchFromGitHub }:
```
sed is part of the stdenv.

File: pkgs/development/python-modules/pyyaml-include/default.nix [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "2343c4dad744d3ce907ec50683b29b4383c7c967f142275bdad8ed56d4de9d94";
+  };
+
+  nativeBuildInputs = [
+    pyyaml
+    setuptools-scm
+    setuptools-scm-git-archive
+    toml
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pyyaml-include/default.nix:22]: That should be part of setuptools-scm actually.

File: pkgs/development/python-modules/plumbum/default.nix [hunk 10]
```diff
@@ -1,19 +1,64 @@
...
+  ];
 
-  src = fetchPypi {
-    inherit pname version;
-    sha256 = "317744342c755319907c773cc87c3a30adaa3a41b0d34c0ce02d9d1904922dce";
+  meta = with lib; {
+    description = "Shell combinators";
+    homepage = "https://github.com/tomerfiliba/plumbum";
+    license = licenses.mit;
+    maintainers = [ ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/plumbum/default.nix:62]: The contributing guide wants us to set this.

File: pkgs/development/python-modules/plumbum/default.nix [hunk 11]
```diff
@@ -1,19 +1,64 @@
...
+  SETUPTOOLS_SCM_PRETEND_VERSION = version;
+
+  nativeBuildInputs = [
+    setuptools-scm
+  ];
+
+  checkInputs = [
+    paramiko
+    psutil
+    pytest-cov
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/plumbum/default.nix:33]:
```suggestion
```
collecting coverage data has no benefit for us or upstream except that it takes ressources.

**Comment by samuelludwig** [comment]
Unfortunately it seems the main functionality of this package is broken, in a similar way as to when I made my attempt.

`${copier} --help` prints whats expected correctly, however, beyond just the help message, little works as it ought to:

If you set up a simple template like so:

```bash
mkdir my-template
cd my-template
echo "filename:
  type: str
  help: What's the name of the file?
  default: myDefaultName
  " > copier.yml
echo "Hi, I'm [[ filename ]]" > '[[ filename ]].txt.tmpl'
```

and then run `copier my-template cop-test`

you should get a prompt requesting a value for `filename`, supposing we entered "working", in `cop-test` there should then be one file named `working.txt` with the contents "`Hi, I'm working`".

When _this_ package is used, versus the copier installed via `pipx`, a similar (but not identical) prompt appears, and seems to accept a value, however you will find that only the raw file is copied over, with no substitutions done. So in `cop-test` there will be a file named `'[[ filename ]].txt.tmpl'` with the contents "`Hi, I'm [[ filename ]]`".

**Comment by jonringer** [comment]
> When this package is used, versus the copier installed via pipx, a similar (but not identical) prompt appears, and seems to accept a value, however you will find that only the raw file is copied over, with no substitutions done. So in cop-test there will be a file named '[[ filename ]].txt.tmpl' with the contents "Hi, I'm [[ filename ]]".

This seems to be intended behavior https://github.com/copier-org/copier/blob/master/CHANGELOG.md#changed 
> We use Jinja 2 defaults now. {{ }} instead of [[ ]] and similar.

One thing to note is that I'm packaging off their master since it's more coherent with the rest of python packages. Breaking changes won't be reflected.

**Comment by samuelludwig** [comment]
> 
> 
> > When this package is used, versus the copier installed via pipx, a similar (but not identical) prompt appears, and seems to accept a value, however you will find that only the raw file is copied over, with no substitutions done. So in cop-test there will be a file named '[[ filename ]].txt.tmpl' with the contents "Hi, I'm [[ filename ]]".
> 
> This seems to be intended behavior https://github.com/copier-org/copier/blob/master/CHANGELOG.md#changed
> 
> > We use Jinja 2 defaults now. {{ }} instead of [[ ]] and similar.
> 
> One thing to note is that I'm packaging off their master since it's more coherent with the rest of python packages. Breaking changes won't be reflected.

oh no... that means i might have had it in an actual working state before... my apologies, thank you, I at least for sure didn't properly package any of the dependencies for nix as you did here

File: pkgs/top-level/python-packages.nix [hunk 12]
```diff
@@ -4792,6 +4792,8 @@ in {
 
   mkdocs-material-extensions = callPackage ../development/python-modules/mkdocs-material-extensions { };
 
+  mkdocs-mermaid2-plugin = callPackage ../development/python-modules/mkdocs2-mermaid2-plugin { };
```
Comment (fbergroth) [inline pkgs/top-level/python-packages.nix:4797]: `mkdocs2` in `mkdocs2-mermaid2-plugin` looks like a typo

**Comment by hhoeflin** [comment]
Would be great if this could be merged! Thanks

**Comment by jonringer** [comment]
related issue: https://github.com/mtkennerly/poetry-dynamic-versioning/issues/62

File: pkgs/tools/misc/copier/default.nix [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+python3.pkgs.buildPythonApplication rec {
+  pname = "copier";
+  version = "unstable-2021-11-10";
```
Comment (yajo) [inline pkgs/tools/misc/copier/default.nix:5]: FWIW [Copier 6 was recently released](https://github.com/copier-org/copier/releases/tag/v6.0.0). You'd probably prefer to package that?

**Pull request closed without merging**